File: 2019-11/dnd-character/python/dnd_character.py
import random, math

class Character:
    def __init__(self):
        self.strength = self.ability()
        self.dexterity = self.ability()
        self.constitution = self.ability()
        self.intelligence = self.ability()
        self.wisdom = self.ability()
        self.charisma = self.ability()
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        dice_list = [random.randint(1,6) for dice in range(1,5)]
        return sum(sorted(dice_list)[1:4])

def modifier(score):
    return math.floor((score-10)/2)

File: 2019-11/dnd-character/javascript/dnd-character.js
//
// This is only a SKELETON file for the 'DnD Character' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const abilityModifier = () => {
  throw new Error("Remove this statement and implement this function");
};

export class Character {
  strength = rollAbility();

  static rollAbility() {
    let array = [];

    for (let i = 0; i < 4; i++) {
      array.push(Math.floor(Math.random() * 6));
    }

    array.sort((a, b) => a - b);
    array.shift();

    const val = array.reduce((a, b) => a + b);

    if (val > 18) {
      throw new Error('Ability scores can be at most 18')
    } else {
      return val;
    }
  }

  get strength() {
    return this.strength;
  }

  get dexterity() {
    return Character.rollAbility();
  }

  get constitution() {
    return Character.rollAbility();
  }

  get intelligence() {
    return Character.rollAbility();
  }

  get wisdom() {
    return Character.rollAbility();
  }

  get charisma() {
    return Character.rollAbility();
  }

  get hitpoints() {
    throw new Error("Remove this statement and implement this function");
  }
}
